relay-general/src/store/transactions.rs

```diff
@@ -705,14 +713,36 @@ mod tests {
             ..Default::default()
         });
 
-        assert_eq_dbg!(
-            process_value(
-                &mut event,
-                &mut TransactionsProcessor::new(None),
-                ProcessingState::root()
-            ),
-            Err(ProcessingAction::InvalidTransaction("span is missing op"))
-        );
+        process_value(
+            &mut event,
+            &mut TransactionsProcessor::new(None),
```

Comment:
You need to set `processor.now` here to avoid clock drift correction (see other tests):

```rs
        let mut processor = TransactionsProcessor::new(None);
        processor.now = end;
        process_value(&mut event, &mut processor, ProcessingState::root()).unwrap();
```

@untitaker clock drift correction looks extremely fragile. It should use a timestamp of when the event was received, not when the processor runs. Since we can have arbitrary delays inside Relay due to project config fetching, this can massively skew the clock -- more than what's expected from usual clock skew. This means, the `TransactionsProcessor` should always receive `sent_at` and also `received_at`.

Also, doing clock drift correction with the `end` timestamp probably does more harm than it helps. We cannot assume that the transaction is sent immediately after it's complete (especially if we want to start batching up in the SDK).

Reply:
Your comments wrt capturing `now` outside of the processor and inside of the endpoint make sense.

>We cannot assume that the transaction is sent immediately after it's complete (especially if we want to start batching up in the SDK).

But that's precisely why the `sent_at` parameter exists. It is eventually supposed to be passed by the SDK (we didn't impl that part yet), and is supposed to tell Relay what the SDK thinks is the current time. The difference between `sent_at` and `now/received_at` is our skew.

Reply:
@jan-auer thanks for pointing out the problem. Dubious default behavior. And surprising API, setting `processor.now` to disable clock drift correction.

The documentation of `TransactionsProcessor.now` reads:

```rust
// This is an attribute so we can mock it in testing.
```

Nothing about clock drift. `TransactionsProcessor` documentation also does no mention it's surprising behavior of modifying input timestamps.

I've cut https://github.com/getsentry/relay/issues/535 to track improving/documenting that.


---

As for the discussion around the clock drift implementation, I've cut https://github.com/getsentry/relay/issues/534 so we can move the discussion there.